Ace

jessebuitenhuis opened the pull request.

Commit:
Added header component

Commit:
Added weight component and routing for it

Commit:
Added progress module with routing

Commit:
Fixed bug with empty routerlink active class

- Fixed issue where starting/empty router was remaining as active. To solve this added routerLinkActiveOptions of exact: true

Commit:
Added base components for weight management

- Added weight service
- Added goal value stored in service to display and manipulate it
- Added weight goal form component
- Added logic that stores weight goal

Commit:
Added weight upsert form fields and emitter

- Added weight model (interface)

Commit:
Added added weight list

- Added subject that holds all added weight values
- Added display of added weight list to weight progress component

Commit:
Added solution for displaying remaining value until goal

- Changed solution how current goal is stored
- Added OnPush change detection strategy
- Added weight progress interface

Commit:
Fixed minor issue with progress not rendering if record was added after goal

Commit:
Added display of previously set goal after route change

jessebuitenhuis commented on src/app/app-routing.module.ts:
```diff
@@ -1,7 +1,21 @@
+    path: '',
+    component: WeightComponent
+  },
+  {
+    path: 'progress',
+    loadChildren: () => import('./modules/progress/progress.module').then(m => m.ProgressModule)
+  },
```
Nice, lazy loading! 👍

jessebuitenhuis commented on src/app/layout/header/header.component.scss:
```diff
@@ -0,0 +1,6 @@
+.navbar {
+    transition: 0.3s ease;
+    .active {
+        font-weight: 500;
+    }
+}
```
Cool, you use nested scss syntax.

jessebuitenhuis commented on src/app/app.component.html:
```diff
@@ -1 +1,2 @@
-<h1>Dialog Assesment (FE)</h1>
+<app-header></app-header>
```
Great job extracting related code to seperate component.

jessebuitenhuis commented on src/app/modules/weight/components/weight-goal-form/weight-goal-form.component.html:
```diff
@@ -0,0 +1,9 @@
+<form class="d-flex flex-row align-items-center">
+  <div class="form-group w-100 me-2">
+    <label for="goal">Goal weight:</label>
+    <input type="number" class="form-control" step="0.01" id="goal" name="goal"
+      [formControl]="goalControl" [class.is-invalid]="goalControl.invalid">
```
Using ReactiveForms over TemplateDrivenForms, like it!

jessebuitenhuis commented on src/app/modules/weight/components/weight-goal-form/weight-goal-form.component.ts:
```diff
@@ -0,0 +1,32 @@
+  selector: 'app-weight-goal-form',
+  templateUrl: './weight-goal-form.component.html',
+  styleUrls: ['./weight-goal-form.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
```
Like it!

jessebuitenhuis commented on src/app/modules/weight/components/weight-goal-form/weight-goal-form.component.ts:
```diff
@@ -0,0 +1,32 @@
+      .pipe(
+        // Making ssure goal is truthy and with valid value
+        filter(goal => goal && goal >= 0),
+        first()
```
Great job not forgetting to unsubscribe and thuis preventing a memory leak.

jessebuitenhuis commented on src/app/modules/weight/components/weight-progress/weight-progress.component.ts:
```diff
@@ -0,0 +1,46 @@
+    private readonly weightService: WeightService,
+    private cdr: ChangeDetectorRef
+  ) {
+    combineLatest([
+      this.weight$,
+      this.weightService.getGoal()
+    ])
+      .pipe(
+        filter(([weight, goal]) => weight.length && goal >= 0),
+        takeUntil(this.unsubscribe$)
+      ).subscribe(([weight, goal]) => {
+        const latestWeight = weight[weight.length - 1].weight;
+        const value = latestWeight - goal;
+        const percent = goal / latestWeight * 100;
+
+        this.remaining = { value, percent };
+        // Marking for check because `remaining` variable is used in view
+        // and we need to render this value. There are no other triggers that
+        // trigger change detection and we set this value only when there is goal and
+        // we have weight records
+        this.cdr.markForCheck();
+      });
```
nested logic, in constructor. Fine for now, would be a review point in our codebase :-)

6matko replied:
Yeah. Tbh not really happy with whole solution but since it was the last piece of work during this task then time pressure was upon and decided to go with the "let it work" approach. If you don't mind then we can discuss this approach and look after different one/better one during our pair programming session.

jessebuitenhuis commented on src/app/modules/weight/components/weight-progress/weight-progress.component.ts:
```diff
@@ -0,0 +1,46 @@
+        const percent = goal / latestWeight * 100;
+
+        this.remaining = { value, percent };
+        // Marking for check because `remaining` variable is used in view
+        // and we need to render this value. There are no other triggers that
+        // trigger change detection and we set this value only when there is goal and
+        // we have weight records
+        this.cdr.markForCheck();
```
Right. Do you do this manual optimization a lot?

6matko replied:
You mean manual optimization with `OnPush` strategy or with `markForCheck` / `detectChanges` in general ? If `OnPush` then I try to use it as much as I can to be in control of change detection and provide better performance because its not really that hard. I don't know how much of a difference it would be in a large project. If its more about `markForCheck` then I use it only when I see that something is not triggering the detection and thus I need to manually do it. Hope I've answered your question :)

jessebuitenhuis commented on src/app/modules/weight/components/weight-progress/weight-progress.component.ts:
```diff
@@ -0,0 +1,46 @@
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.unsubscribe$.next();
+    this.unsubscribe$.complete();
+  }
```
Cool! We've added a helper for this to not repeat yourself doing this exactly in our codebase.

6matko replied:
I know about this one as well: https://github.com/ngneat/until-destroy - Used in one of my pet projects.

jessebuitenhuis commented on src/app/modules/weight/components/weight-upsert-form/weight-upsert-form.component.html:
```diff
@@ -0,0 +1,15 @@
+  </div>
+  <div class="form-group w-100 me-2">
+    <label for="comment">Comment:</label>
+    <textarea class="form-control" id="comment" placeholder="Had trash food"
+      formControlName="comment"></textarea>
```
Inspiring placeholder 😆

jessebuitenhuis commented on src/app/modules/weight/components/weight-upsert-form/weight-upsert-form.component.html:
```diff
@@ -0,0 +1,15 @@
+<form class="mt-4 card p-4" [formGroup]="form">
+  <div class="form-group w-100 me-2">
+    <label for="weight">Weight:</label>
+    <input type="number" class="form-control" step="0.01" id="weight"
+      name="weight"
+      formControlName="weight" [class.is-invalid]="form.get('weight').invalid">
```
Super, added validation!

jessebuitenhuis commented on src/app/modules/weight/weight.service.ts:
```diff
@@ -0,0 +1,33 @@
+  providedIn: 'root'
+})
+export class WeightService {
+  /**
+   * Stored weight goal for reusability during this task. Otherwise should
+   * be stored in DB
+   *
+   */
+  private goalSubject = new BehaviorSubject<number>(null);
+  private weightSubject = new BehaviorSubject<Weight[]>([]);
```
Perfect solution for this assignment!

jessebuitenhuis reviewed (commented):
Great job on the assignment! I've added some comments right away, so you can see a little bit how we would normally handle Pull Requests and Code Reviews in our codebase.

Commit:
Added first solution for progress estimation feature

- Added a few unit tests regarding weight progress component
- Added calculation methods
- Added a few comments
- Added display of goal estimation
* TODO: Cleanup code and refactor

Commit:
Added a few comments and extracted remaining weight progress to separate method

Commit:
Moved combineLatest logic to separate method and added a few unit tests

- Moved logic that was responsible for actions when weight records & goals change to separate method in order to unit test. Previously it was kinda hard to do without it
- Added more unit tests to test method logic and if remaining / required estimation is calculated

Commit:
Removed extra type guard and made code a little bit simpler

Commit:
Removed redundant comment
